### DISCUSSION TITLE
Inject custom user Identity by typing

### DISCUSSION THREAD

**Discussion by nkhitrov** [Neoteroi/BlackSheep#491] [Ideas]
I wrote a custom authentication handler with custom Identity to use property with typing:
```python
class MyUserIdentity(Identity):

    @property
    def id(self) -> int:
        return self.get("id")

    @property
    def username(self) -> str:
        return self.get("username")
```

```python
class JWTAuthenticationHandler(AuthenticationHandler):

    def __init__(self, ...):
		...

    async def authenticate(self, reqeust: Request) -> MyUserIdentity | None:
		...
```

When I set `MyUserIdentity` as typing for user param It does not work.
When I use `guardpost.User` as typing for user, `blacksheep` injects my custom identity instance, but in this case controller has invalid typing.

```python
from guardpost import User

...
@router.get("/users/me")
async def get_me(self: Self, user: User | None) -> Response:
        print(user)  #  instance of MyUserIdentity

        if user:
            return self.json({"id": user.id, "username": user.username})

        return self.json({"id": None})
```

I think we should change [this line](https://github.com/Neoteroi/BlackSheep/blob/026897ff2796c2f2074cb368ff9c59a262dc28ad/blacksheep/server/normalization.py#L427) from `if annotation is User or annotation is Identity` to `if issubclass(annotation, Identity)`

**Comment by tyzhnenko** [comment]
@nkhitrov take a look @RobertoPrevato's examples in this discussion https://github.com/Neoteroi/BlackSheep/discussions/338. I hope they will be useful.

I use a custom Identity class

```python
class UserIdentity(Identity):
    @property
    def id(self) -> Optional[str]:
        return self["id"] or self.sub

    @property
    def session_id(self) -> Optional[str]:
        return self["sid"] or None

    @property
    def scopes(self) -> list[str]:
        return self["scopes"]
```

I set an instance of this class to `request.identity` in auth middleware and then get it via the factory

```python
    request.identity = UserIdentity(token, self.token_type.decode())
```

```
@add_service(scope="scoped_factory")
def auth_user_factory(context) -> UserIdentity:
    request: Request = context.scoped_services["__request__"]
    return request.identity  # type: ignore
```
